sql/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala

```diff
@@ -3723,7 +3728,47 @@ class SparkConnectPlanner(
     LazyExpression(transformExpression(getLazyExpression.getChild))
   }
 
-  private def assertPlan(assertion: Boolean, message: String = ""): Unit = {
+  private def transformSubqueryExpression(
+      getSubqueryExpression: proto.SubqueryExpression): Expression = {
+    val planId = getSubqueryExpression.getPlanId
+    getSubqueryExpression.getSubqueryType match {
+      case proto.SubqueryExpression.SubqueryType.SUBQUERY_TYPE_SCALAR =>
+        UnresolvedScalarSubqueryPlanId(planId)
+      case proto.SubqueryExpression.SubqueryType.SUBQUERY_TYPE_EXISTS =>
+        UnresolvedExistsPlanId(planId)
+      case other => throw InvalidPlanInput(s"Unknown SubqueryType $other")
+    }
+  }
+
+  private def transformWithRelations(getWithRelations: proto.WithRelations): LogicalPlan = {
+    // Wrap the plan to keep the original planId.
+    val plan = Project(Seq(UnresolvedStar(None)), transformRelation(getWithRelations.getRoot))
+
+    val relations = getWithRelations.getReferencesList.asScala.map {
+      case ref if ref.hasCommon && ref.getCommon.hasPlanId =>
+        val planId = ref.getCommon.getPlanId
+        val plan = transformRelation(ref)
+        planId -> plan
+      case _ => throw InvalidPlanInput("Invalid WithRelation reference")
+    }.toMap
+
+    val withRelations = plan
+      .transformAllExpressionsWithPruning(_.containsPattern(TreePattern.UNRESOLVED_PLAN_ID)) {
+        case u: UnresolvedPlanId if relations.contains(u.planId) =>
```

Comment:
You could also collect the missing plan ids in this step. That saves you from traversing the tree twice.

Reply:
ah, that sounds good!